README.md
# Adsorption Energy of Any Molecule on Metals Using Graph Neural Networks

<p align="center">
    <img src="./GNN.gif" width="60%" height="60%"/>
</p>

This is the official repository of the framework related to the article "Adsorption Energy of Any Molecule on Metals Using Graph Neural Networks". The Graph Neural Networks (GNNs) developed within this framework allow the fast prediction of the DFT ground state energy of the following systems:

- All gas-phase closed-shell molecules containing C, H, O, N and S.
- Adsorption systems: Same molecules mentioned above on the following 14 metals: Ag, Au, Cd, Co, Cu, Fe, Ir, Ni, Os, Pd, Pt, Rh, Ru, Zn.

The framework is built on top of [PyTorch](https://pytorch.org/) and [PyTorch Geometric](https://pytorch-geometric.readthedocs.io/en/latest/index.html).
## Installation

1. Clone the repo from GitLab. Open a terminal and use the following command:  

    `git clone https://gitlab.com/iciq-tcc/nlopez-group/gnn_eads.git`  
    
    You should now have the repo ``gnn_eads`` in your current directory.  

2. Create a conda environment. Enter the repo: You should find the file `GNN_env.txt`: It contains all the information about the packages needed to create the environment for the GNN framework. Choose a name for the new environment (we will use `GNN` here) and type the command:  

    `conda create --name GNN --file GNN_env.txt`  
    
    Check that you have created your new environment by typing `conda env list`: A list with all your environments appear, together with the newly created `GNN`. Activate it with `conda activate GNN` (you will see the name of the current active environment within parentheses on the left of the terminal).  

3. Install pyRDTP, a package at the base of the GNN framework. Clone the repo in the existing `gnn_eads` base directory (i.e., be sure to be in `gnn_eads` before executing the following command) and install it with pip:  

    `git clone --branch experimental https://gitlab.com/iciq-tcc/nlopez-group/pyrdtp.git`  
    `pip install pyrdtp`

    To check the correctness of the installation, type `conda list` and check out the presence of pyrdtp in the list.

4. Create the following empty folders in the repo: 

    `mkdir Data Models Hyperparameter_Optimization`

    As the names suggest, `Data` will store the DFT datasets used to train the GNNs in case you will work in *training mode* (see below), while `Models` will contain the different GNN models stored in [Zenodo](https://www.zenodo.org/) if you choose to work in *inference mode* (see below), or the models you will create if you work in training mode otherwise. `Hyperparameter_Optimization` will store the results of the tuning process.

5. Download the DFT training data from Zenodo: The FG-dataset must be stored in the folder `Data`.

6. Download the GNN models from Zenodo: They must be stored in the folder `Models`.

Done! In theory now everything is set up to start playing with the GNN framework!

## Usage

You have two possible choices for interacting with the GNN framework:

- **Inference mode**: Most likely, you are a curious person and want to probe the performance of the GNN models compared to your DFT simulations. In this case, you will test the pre-trained models developed by us, without going deeper in the details behind the models' generation process and without accessing the DFT data used to train them. 

- **Training mode**: You will go through all the steps defined in the workflow for the GNN model generation process. In this case you will need the DFT datasets for training the GNN. Within this mode, you can train your own models with the preferred hyperparameter setting and model architecture, or, if you have enough computational resources, you can perform hyperparameter tuning with the workflow based on [Ray Tune](https://docs.ray.io/en/latest/tune/index.html).

### Inference mode

Within this mode, you can opt among two different options:

1. You already have performed some DFT calculations with VASP and want to compare the performance of the GNN models with the ground-truth provided by your data. In this case, the main scripts and files you will work with are `GNNvsDFT.ipynb` and `GNNvsDFT.ipynb`.

2. You have no DFT data for a specific system and want to get an estimation from our pre-trained Graph Neural Networks. In this case, you will play with `interactive_graph_creator.py`, an interactive script connected to the [ChemSpider](https://www.chemspider.com) database to help you build the graph related to your specific case, providing the ground state energy of the system and the adjusted adsorption energy. 

    `python interactive_graph_creator.py`
### Training mode

The DFT datasets and the models are stored open-access [here](). The DFT datasets are also uploaded to ioChem-BD [doi]().
Within this mode as well, you can choose among two available ways to work with the GNN:

1. Perform a model training with your own model architecture and hyperparameter setting: To do so, follow the instructions provided in the Jupyter notebook `train_GNN.ipynb`, or directly run the script `train_GNN.py`. For example, to launch a model training called `test` with 200 epochs, a batch size of 16 and Mean Squared Error as loss function, type the following command:

    `python train_GNN.py -o test -e 200 -b 16 -j mse`

    To check the documentation of the script, type `python train_GNN.py -help`.
2. Perform a hyperparameter optimization using the Asynchronous Successive Halving (ASHA) scheduler provided by Ray Tune. You can study the effect of all hyperparameters on the final model performance (e.g., learning rate, loss function, epochs) and you can also test different model architecture automatically, without the need of manually defining the model architecture. The script you have to use in this case is `hypopt_GNN.py`. For this script, the hyperparameter space must be defined in the script before running it. For example, to launch a hyperparameter optimization called `long_night` with 2000 trials, each one with a grace period of 15 epochs and providing 0.5 GPUs for each trial (each GPU partitioned in two), type:

    `python hypopt_GNN.py -o long_night -s 2000 -gr 15 -gpt 0.5`


# Authors

Santiago Morandi, Ph.D. Student, López group at ICIQ (Spain)  
Sergio Pablo-García, Ph.D. Student, López Group (ICIQ, Spain), now postdoc in the Aspuru-Guzik group (UoT, Canada)
# Contributors

Zarko Ivkovic, M.Sc. Student, University of Barcelona, ICIQ Summer Fellow 2022; involved in the creation of the DFT dataset and interface testing.

# License

[MIT License](https://gitlab.com/iciq-tcc/nlopez-group/gnn_eads/-/blob/master/LICENSE)
# Support 

In case you need help or are interested in contributing, feel free to contact us sending an e-mail to smorandi@iciq.es

# Acknowledgements

<p align="center">
    <img src="./ack_repo.png"/>
</p>

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2022-09-06
### Added
- Possibility to include second-order metal atoms in the graph representation.
- Flexible Graph Neural Network base class.


### Changed
- Start using "changelog" over "change log" since it's the common usage.

### Removed



hypopt_GNN.py
"""Run Hyperparameter optimization with Ray Tune"""

__author__ = "Santiago Morandi"

import argparse
import sys
from datetime import date, datetime
import os
os.environ["TUNE_RESULT_DELIM"] = "/"

import torch
from torch.optim.lr_scheduler import ReduceLROnPlateau
from torch_geometric.nn import SAGEConv, GATv2Conv, GraphMultisetTransformer
from torch.nn import ReLU, Tanh, Sigmoid
from torch.optim import Adam
from ray import tune
from ray.tune.schedulers import ASHAScheduler
from ray.tune import Stopper
import ray
import pandas as pd

from constants import NODE_FEATURES
from nets import FlexibleNet
from functions import train_loop, test_loop, scale_target, create_loaders
from processed_datasets import FG_dataset, BM_dataloader
from post_training import create_model_report

HYPERPARAMS = {}
# Process-related
HYPERPARAMS["test_set"] = True          
HYPERPARAMS["splits"] = 10              
HYPERPARAMS["target_scaling"] = "std"   
HYPERPARAMS["batch_size"] = tune.choice([16, 32, 64])           
HYPERPARAMS["epochs"] = 200               
HYPERPARAMS["loss_function"] = torch.nn.functional.l1_loss   
HYPERPARAMS["lr0"] = tune.choice([0.001, 0.0001])       
HYPERPARAMS["patience"] = tune.choice([5, 7])              
HYPERPARAMS["factor"] = tune.choice([0.5, 0.7, 0.9])          
HYPERPARAMS["minlr"] = tune.choice([1e-8])             
HYPERPARAMS["betas"] = (0.9, 0.999)     
HYPERPARAMS["eps"] = tune.choice([1e-8, 1e-9])               
HYPERPARAMS["weight_decay"] = 0         
HYPERPARAMS["amsgrad"] = tune.choice([True, False])       
# Model-related
HYPERPARAMS["dim"] = tune.choice([128, 256])                
HYPERPARAMS["sigma"] = tune.choice([torch.nn.ReLU()])  
HYPERPARAMS["bias"] = tune.choice([True, False])  
HYPERPARAMS["adj_conv"] = tune.choice([True, False])  
HYPERPARAMS["n_linear"] = tune.choice([0, 1, 3, 5]) 
HYPERPARAMS["n_conv"] = tune.choice([3, 4, 5]) 
HYPERPARAMS["conv_layer"] = tune.choice([SAGEConv, GATv2Conv]) 
HYPERPARAMS["pool_layer"] =  tune.choice([GraphMultisetTransformer])           
HYPERPARAMS["conv_normalize"] = False   
HYPERPARAMS["conv_root_weight"] = True
HYPERPARAMS["pool_ratio"] = tune.choice([0.25, 0.5, 0.75])        
HYPERPARAMS["pool_heads"] = tune.choice([1, 2, 4])
HYPERPARAMS["pool_seq"] = tune.choice([["GMPool_I"], 
                                       ["GMPool_G"], 
                                       ["GMPool_G", "GMPool_I"],
                                       ["GMPool_G", "SelfAtt", "GMPool_I"],
                                       ["GMPool_G", "SelfAtt", "SelfAtt", "GMPool_I"]])
HYPERPARAMS["pool_layer_norm"] = False 

    
if __name__ == "__main__":
    PARSER = argparse.ArgumentParser(description="Perform hyperparameter optimization with Ray-Tune with ASynchronous Hyperband Algorithm (ASHA).")
    PARSER.add_argument("-o", "--output", type=str, dest="o", 
                        help="Name of the directory where results will be stored.")
    PARSER.add_argument("-s", "--samples", default=10, type=int, dest="s",
                        help="Number of samples during the search.")
    PARSER.add_argument("-v", "--verbose", default=1, type=int, dest="v",
                        help="Verbosity of tune.run() function")
    PARSER.add_argument("-gr", "--grace", default=15, type=int, dest="grace", 
                        help="Grace period of ASHA.")
    PARSER.add_argument("-maxit", "--max-iterations", default=150, type=int, dest="max_iter", 
                        help="Maximum number of training iterations (epochs) allowed by ASHA.")
    PARSER.add_argument("-rf", "--reduction-factor", default=4, type=int, dest="rf", 
                        help="Reduction factor of ASHA.")
    PARSER.add_argument("-bra", "--brackets", default=1, type=int, dest="bra", 
                        help="Number of brackets of ASHA.")
    PARSER.add_argument("-gpt", "--gpu-per-trial", default=1.0, type=float, dest="gpu_per_trial", 
                        help="Number of gpus per trial (can be fractional).")    
    ARGS = PARSER.parse_args()
        
    def train_function(config, checkpoint_dir=None):
        """
        Function for hyperparameter tuning via RayTune.
        Args:
            config (dict): Dictionary with search space (hyperparameters)
        """    
        train_loader, val_loader, test_loader = create_loaders(FG_dataset,
                                                               config["splits"],
                                                               config["batch_size"], 
                                                               config["test_set"])
        train_loader, val_loader, test_loader, mean, std = scale_target(train_loader,
                                                                        val_loader,
                                                                        test_loader, 
                                                                        mode=config["target_scaling"], 
                                                                        test=config["test_set"])    
        device = "cuda" if torch.cuda.is_available() else "cpu"            
        model = FlexibleNet(dim=config["dim"],
                            N_linear=config["n_linear"], 
                            N_conv=config["n_conv"], 
                            adj_conv=config["adj_conv"], 
                            in_features=NODE_FEATURES, 
                            sigma=config["sigma"], 
                            bias=config["bias"], 
                            conv=config["conv_layer"], 
                            pool=config["pool_layer"], 
                            pool_ratio=config["pool_ratio"], 
                            pool_heads=config["pool_heads"], 
                            pool_seq=config["pool_seq"], 
                            pool_layer_norm=config["pool_layer_norm"]).to(device)     
        optimizer = Adam(model.parameters(),
                         lr=config["lr0"], 
                         betas=config["betas"],
                         eps=config["eps"], 
                         weight_decay=config["weight_decay"], 
                         amsgrad=config["amsgrad"])
        lr_scheduler = ReduceLROnPlateau(optimizer,
                                         mode='min',
                                         factor=config["factor"],
                                         patience=config["patience"],
                                         min_lr=config["minlr"]) 
    
        train_list = [] # Store training MAE during training
        val_list = []   # Store validation MAE during training
        test_list = []  # Store test MAE during training   
        for iteration in range(1, config["epochs"]+1):
            lr = lr_scheduler.optimizer.param_groups[0]['lr']
            _, train_MAE = train_loop(model, device, train_loader, optimizer, config["loss_function"])  
            val_MAE = test_loop(model, val_loader, device, std)
            lr_scheduler.step(val_MAE)  
            if config["test_set"]:
                test_MAE = test_loop(model, test_loader, device, std)                                           
                print('Epoch {:03d}: LR={:.7f}  Train MAE: {:.4f} eV  Validation MAE: {:.4f} eV '             
                      'Test MAE: {:.4f} eV'.format(iteration, lr, train_MAE*std, val_MAE, test_MAE))
            else:
                print('Epoch {:03d}: LR={:.7f}  Train MAE: {:.6f} eV  Validation MAE: {:.6f} eV '
                      .format(iteration, lr, train_MAE*std, val_MAE))      
            BM_MAE = test_loop(model, BM_dataloader, device=device, std=std, mean=mean, scaled_graph_label=False)                    
            tune.report(BM_MAE=BM_MAE, FG_MAE=test_MAE, epoch=iteration)
            
            train_list.append(train_MAE * std)
            val_list.append(val_MAE)
            test_list.append(test_MAE)
        if BM_MAE <= 0.5:  # If the extrapolation performance is good, save the model!
            time = str(datetime.now())[11:]
            time = time[:8]
            create_model_report("{}_{}".format(ARGS.o, time),
                                model,
                                (train_loader, val_loader, test_loader),
                                (mean, std),
                                config,
                                (train_list, val_list, test_list))
        
    hypopt_scheduler = ASHAScheduler(time_attr="epoch",
                                     metric="BM_MAE",
                                     mode="min",
                                     grace_period=ARGS.grace,
                                     reduction_factor=ARGS.rf,
                                     max_t=ARGS.max_iter,
                                     brackets=ARGS.bra)
    ray.init(ignore_reinit_error=True)
    result = tune.run(train_function,
                        name=ARGS.o,
                        time_budget_s=3600*24,
                        config=HYPERPARAMS,
                        scheduler=hypopt_scheduler,
                        resources_per_trial={"cpu":1, "gpu":ARGS.gpu_per_trial},
                        num_samples=ARGS.s, 
                        verbose=ARGS.v,
                        log_to_file=True, 
                        local_dir="./Hyperparameter_Optimization",
                        raise_on_failed_trial=False)
    best_config = result.get_best_config(metric="BM_MAE", mode="min", scope="last")
    best_config_df = pd.DataFrame.from_dict(best_config, orient="index")    
    best_config_df.to_csv("./Hyperparameter_Optimization/{}/best_config.csv".format(ARGS.o), sep="/")
    print(best_config)
    exp_df = result.results_df
    exp_df.to_csv("./Hyperparameter_Optimization/{}/summary.csv".format(ARGS.o), sep="/")

    
